File: fx/src/main/java/gargoyle/fx/icons/Icon.java
package gargoyle.fx.icons;

import gargoyle.fx.FXConstants;
import gargoyle.fx.FXContext;
import gargoyle.fx.FXUtil;
import javafx.scene.control.ButtonBar;

import java.net.URL;
import java.util.Arrays;
import java.util.Locale;
import java.util.Optional;

import static javafx.scene.control.ButtonBar.ButtonData.CANCEL_CLOSE;
import static javafx.scene.control.ButtonBar.ButtonData.NO;
import static javafx.scene.control.ButtonBar.ButtonData.OK_DONE;
import static javafx.scene.control.ButtonBar.ButtonData.YES;

public enum Icon {
    CANCEL("cancel", CANCEL_CLOSE, NO),
    EMPTY("empty"),
    OK("ok", OK_DONE, YES);

    private final String value;
    private final ButtonBar.ButtonData[] buttonData;

    Icon(final String value, final ButtonBar.ButtonData... buttonData) {
        this.value = value;
        this.buttonData = buttonData;
    }

    public static Icon find(final ButtonBar.ButtonData buttonData) {
        return Arrays.stream(values()).filter(icon ->
                        Arrays.stream(icon.buttonData).anyMatch(data -> buttonData == data)).findFirst()
                .orElse(EMPTY);
    }

    public Optional<URL> findIcon(final FXContext context) {
        final String iconBaseName = FXUtil.resolveBaseName(Icon.class, value);
        final Class<?> baseClass = context.getBaseClass();
        final Locale locale = context.getLocale();
        return FXUtil.findResource(Icon.class, locale, iconBaseName, FXConstants.EXT__IMAGES).or(() ->
                FXUtil.findResource(baseClass, locale, iconBaseName, FXConstants.EXT__IMAGES));
    }

}

File: fx/src/main/kotlin/gargoyle/fx/icons/Icon.kt
package gargoyle.fx.icons

import gargoyle.fx.FxConstants
import gargoyle.fx.FxContext
import gargoyle.fx.FxUtil
import javafx.scene.control.ButtonBar.ButtonData
import java.net.URL

enum class Icon(private val value: String, vararg buttonData: ButtonData) {
    CANCEL("cancel", ButtonData.CANCEL_CLOSE, ButtonData.NO),
    EMPTY("empty"),
    OK("ok", ButtonData.OK_DONE, ButtonData.YES);

    private val buttonData: Array<ButtonData> = arrayOf(*buttonData)

    fun findIcon(context: FxContext): URL? =
        FxUtil.resolveBaseName(Icon::class, value).let {
            FxUtil.findResource(Icon::class, context.locale, it, FxConstants.EXT__IMAGES)
                ?: FxUtil.findResource(context.baseClass, context.locale, it, FxConstants.EXT__IMAGES)
        }

    companion object {
        fun find(buttonData: ButtonData): Icon =
            values().firstOrNull { icon -> icon.buttonData.any { buttonData == it } } ?: EMPTY
    }
}
